java-checks/src/main/java/org/sonar/java/checks/CognitiveComplexityMethodCheck.java

```diff
@@ -0,0 +1,476 @@
+
+    if (binTree.is(CONDITIONAL_AND)) {
+      while (binTree.leftOperand().is(CONDITIONAL_AND)) {
+        binTree = (BinaryExpressionTree) binTree.leftOperand();
+      }
+    } else if (binTree.is(CONDITIONAL_OR)) {
+      ExpressionTree left = binTree.leftOperand();
+      while (left.is(CONDITIONAL_OR) && ! isLogicalOp(((BinaryExpressionTree)left).rightOperand())) {
+        binTree = (BinaryExpressionTree) binTree.leftOperand();
+        left = binTree.leftOperand();
+      }
+    }
+    return binTree;
+  }
+
+
+  private boolean isLogicalOp(ExpressionTree expTree) {
+    return expTree.is(CONDITIONAL_AND) || expTree.is(CONDITIONAL_OR);
+  }
+
+  private int countTryChains(Tree st, List<JavaFileScannerContext.Location> flow, int nestingLevel) {
+    int total = 0;
+
+    if (st.is(Tree.Kind.TRY_STATEMENT)) {
+      TryStatementTree tryStatment = (TryStatementTree) st;
+
+      List<CatchTree> catches = tryStatment.catches();
+      for (CatchTree catchTree : catches) {
+        total++;
+        total += nestingLevel;
+        addSecondaryLocation(flow, catchTree, total, nestingLevel);
+        total += countComplexity(catchTree.block().body(), flow, nestingLevel +1);
+      }
+
+      if (tryStatment.finallyBlock() != null && tryStatment.finallyBlock().body() != null) {
+        total += countComplexity(tryStatment.finallyBlock().body(), flow, nestingLevel);
+      }
+    }
+
+    return total;
+  }
+
+  private int countBreakAndContinue(Tree st, List<JavaFileScannerContext.Location> flow) {
+    if ( (st.is(BREAK_STATEMENT) && ((BreakStatementTree) st).label() != null)
+            || (st.is(CONTINUE_STATEMENT) && ((ContinueStatementTree) st).label() != null) ) {
+      flow.add(new JavaFileScannerContext.Location("+1", st));
+      return 1;
+
+    }
+
+    return 0;
+  }
+
+  private Tree getSegmentOfInterest(StatementTree st) {
+    switch (st.kind()) {
+      case LABELED_STATEMENT:
+        return ((LabeledStatementTree)st).statement();
+      case VARIABLE:
+        VariableTree vt = (VariableTree)st;
+        if (vt.initializer() != null) {
+          return vt.initializer();
+        }
+        break;
+      case EXPRESSION_STATEMENT:
+        if (((ExpressionStatementTree)st).expression() != null && ((ExpressionStatementTree)st).expression().is(Tree.Kind.ASSIGNMENT)) {
+          AssignmentExpressionTree aet = (AssignmentExpressionTree)((ExpressionStatementTree) st).expression();
+          if (aet.expression() != null && aet.expression().is(Tree.Kind.NEW_CLASS) && ((NewClassTree) aet.expression()).classBody() != null) {
+            return ((NewClassTree) aet.expression()).classBody();
+          }
+        }
+        break;
+      case RETURN_STATEMENT:
+        if (((ReturnStatementTree) st).expression() != null) {
+          return ((ReturnStatementTree) st).expression();
+        }
+        break;
+      default:
+        break;
+    }
+
+    return st;
+  }
+
+
+  private List<StatementTree> getChildren(Tree st) {
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) The Cyclomatic Complexity of this method "getChildren" is 20 which is greater than 10 authorized. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Make "getChildren" a "static" method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AS2325)